### ISSUE TITLE
Possible bug in StringValidationRule.getValid  with canonicalize

### ISSUE THREAD

**Issue by DavidSmithAvaloq** [ESAPI/esapi-java-legacy#588]
Using version 2.2.1.0
In StringValidationRule.getValid  with canonicalize set I see

```java
		// canonicalize
		if (canonicalizeInput) {
		    data = encoder.canonicalize(input);
		} else {
		    String message = String.format("Input validaiton excludes canonicalization.  Context: %s   Input: %s", context, input);
		    LOGGER.warning(Logger.SECURITY_AUDIT, message);
            data = input;
		}

		// check whitelist patterns
		checkWhitelist(context, input);

		// check blacklist patterns
		checkBlacklist(context, input);
```

This sets data to be the canonical version of the string but then appears to do the validation on the original input string.

**Comment by xeno6696** [comment]
@DavidSmithAvaloq looking at where the method gets called, it's being used in `Validator.getValidInput()`, as well as from `BaseValidationRule`, `CreditCardValidationRule`, and `HTMLValidation` rule.  

That part of the API is designed to assert that the input data is untrusted with an idiomatic approach that suggests the user intends on receiving 'safe' data from an untrusted source.  

@planetlevel -->  Do you by any chance recall conversations as to why we chose to return altered data as opposed to original data that we reasonably assert is safe?  (I.E. --> running canonicalize to detect mixed/multi encoding but ultimately returning the original data?)

**Comment by planetlevel** [comment]
There were some very vocal (if not helpful) members of the community that were loudly against modifying data. I think they thought this would introduce some sort of integrity problem. I disagreed and still disagree. The safe behavior should be the default, and if someone wants to specifically get the unsafe data, it should be possible -- like getOriginalAndPossiblyUnsafeInput() or something.

**Comment by xeno6696** [comment]
@planetlevel  While I think there's room for nuance especially with some aggressive request validation frameworks I won't resurrect old battles and I have a general penchant to NOT change something especially if its as fundamental and longstanding as this bit of code.  

I'm going to close this issue.  For my part @DavidSmithAvaloq I agree with your approach, but given that the philosophy governing this part of the library was built with the opposite intent I'm going to suggest a simple workaround:  If you want to store the original input data, make two calls, one with and one without canonicalize.  You _will_ double the validation workload but given the regexes we use are almost always length-limited this shouldn't be an issue.  Of whitelist/blacklist/canonicalize canonicalize is the most expensive method and would only get called once here.   

If you would like to add a method similar to what Jeff suggested here, by all means fork and submit a PR.  

To be crystal clear to anyone who reads this in the future, this behavior _is by design_.

**Comment by DavidSmithAvaloq** [comment]
Perhaps my initial report was not clear enough. My concern is not with what gets returned.
My problem is that we do this `data = encoder.canonicalize(input)` but then we do this `checkWhitelist(context, input);`

That is, we do the **validation** on the original input so what is the point of calling canonicalize?

**Comment by kwwall** [comment]
@DavidSmithAvaloq In my opinion, this is a bug. Whether or not is should be fixed I suppose is open to debate because it could change compatibility with previous ESAPI versions, but if we are careful and don't mind introducing a bit of additional complexity via a new property in ESAPI.properties, we could have it both ways, which might be important for some. If we don't fix it, I think this behavior should at least be noted in the ESAPI Javadoc. I am discussing options with my colleagues off-list. If we decide to fix it, we will reopen this.

**Comment by DavidSmithAvaloq** [comment]
@kwwall Thanks for still looking at this. I think that this behavior changed a while ago and it was when I updated the library I use I saw the modified behavior. Back at (for instance) esapi-2.1.0.1 we called checkWhitelist with **both** strings and validated against the canonical one:
```
// check whitelist patterns
		checkWhitelist(context, data, input);
```

**Comment by xeno6696** [comment]
> Perhaps my initial report was not clear enough. My concern is not with what gets returned.
> My problem is that we do this `data = encoder.canonicalize(input)` but then we do this `checkWhitelist(context, input);`
> 
> That is, we do the **validation** on the original input so what is the point of calling canonicalize?

Okay yes, that IS a problem and I missed that.  Reopening!

**Comment by xeno6696** [comment]
I see what I did.  In the diff between what I committed in 2017 and the previous commit, I made the following changes:

1.  refactored the length check to happen irregardless of whether or not we wish to canonicalize (fast-fail;  why bother doing anything else if the length is blown?
2.  I noticed that in the original code we have this:  
![image](https://user-images.githubusercontent.com/9502785/106776529-63c83e00-6601-11eb-9b60-1e838b529ca8.png)
Notice that in the original implementation we're doing regex validation against the original input... and **not** the canonicalized version.  The current version is this:

![image](https://user-images.githubusercontent.com/9502785/106776816-a4c05280-6601-11eb-957f-d48aca0bb15e.png)

For both the checkWhitelist and checkBlacklist implementations we have the following signature:

`private String checkWhitelist(String context, String input, String orig)`

So I think this could be a bad variable naming issue.  

`data = encoder.canonicalize(input);`

(line 275)
```
//sig:  checkWhitelist(String context, String input, String orig)
checkWhitelist(context, data, input);
//sig:  checkBlacklist(String context, String input, String orig)
checkBlacklist(context, data, input);
```
So, we receive input, and then check the canonicalized form.  "data" should maybe be renamed "canonicalized" but I seem to have corrected a longstanding bug in the original code.  @kwwall if you want I can change this back to the old way, where we validate against the original input and canonicalize last?  (But then... kinda defeat the idea of canonicalize in the first place?)

**Comment by xeno6696** [comment]
So the old order was this:
1. If we want to canonicalize
2. Check length
3. check whitelist 
4. check blacklist
5. canonicalize
6. check length
7. check whitelist
8. check blacklist

I reduced this to:

1.  Check length
2. if we want to canonicalize
3. canonicalize
4. check whitelist
5. check blacklist

I don't have my notes from 2017 anymore, but my opinion today is this:

Original 3:  why before canonicalize? 
Original 4:  why before canonicalize?
Original 6:  We'd check length after canonicalization for what purpose?  In all cases if html/javascript/uri encodings are discovered (singly) they will be reduced to their single-character representation.  (i.e. %5C = '\').  This second check seems superfluous and guaranteed to always be true if the original input passes this check. The only time this would be false is if there was a bug in our own code causing a decode function to act like an encode function and I don't think it would be prudent to write code to check _here_ for problems _there_.

After canonicalization we can be reasonably sure of the following
a.  No mixed encoding is present
b.  No multi-encoding is present
c.  We will have an input that represents the final String.  

I'm thinking that Original 3&4 were there to fail-fast in cases where the inputs were just wholly illegal in the first place.  (Not security threats, but just plain wrong.)  this optimizes I guess by not performing canonicalization, however this _feels wrong_
 as I've **never** validated something without first ensuring it's byte encoding was right for my platform.    In the Java context this means UTF-16 is already decided for us, so then we just need to make sure that the string represents a canonicalized form.  

I don't think the original logic makes much sense.  BUT since @DavidSmithAvaloq seems to have detected an anomaly, could you share the input that is causing the issue?  Or are you just wondering about the impacts of the code change?

**Comment by DavidSmithAvaloq** [comment]
We have a unit test which failed when I tried to upgrade to the newer version of the library. esapi is not something I use myself (I just happened to have the task of updating some out-of-date libraries). As far as I can tell, we have this unit test:

`assertFalse(this.stringParamValidator.isValid("&quot;test&quot;", this.context));`

which works with 2.1.0.1 (that is, the input was rejected) but fails with the 2.2.1.1 (that is, the input gets accepted when  we expect it to be rejected). The unit test effectively I think) does this:

```
Validator.avaloqLooseSafeString=^[+>=<?!%_,~:@|\\-\\/\\*\\&\\;\\.\\#\\n\\r\\p{Pd}\\p{Ps}\\p{Pe}\\p{Pc}\\p{N}\\p{L}\\p{Sc}\\p{Z}]{0,10000}$
...
// I don't at the moment know how esapi knows to use the avaloqLooseSafeString, but that appears to be what it uses
...
ESAPI.validator().isValidInput(
        "String validation",    // context
        "&quot;test&quot;",     // input
        "avaloqLooseSafeString",// type
        2147483647,             // maxLength
        true,                   // allowNull
        true);                  // canonicalize
```
I hope that helps, but let me know if you need me to look more into our use of esapi. Or indeed if you need me put together a small runnable example.

**Comment by xeno6696** [comment]
Well... weird.  Off the bat I'm confused:  `&quot;test&quot;` should be valid according to the semantics of that regex.
1.  We explicitly allow `&`
2. We explicitly allow `;`
3. `\p{L}` should take care of `test` and `quot` (as that should cover almost any language text in Unicode that is a letter.)

In taking apart that regex, you have a bug where the minus sign is embedded in the middle of the regex, forcing the engine to think you're creating a range.  I don't think this is a bug per se but I'm already seeing problems with the regex.  
![image](https://user-images.githubusercontent.com/9502785/106908477-94b97900-66bc-11eb-8672-003a0086fdf8.png)

Next, you have the escaped backslash listed multiple times throughout the regex.  I found the bug.

`\\p{L}` is trying to match `\p{L}` instead of the unicode range intended by `\p{L}`

Yep.  Your regex is broken.  it should be:

`^[-+>=<?!%_,~:@|\\/\*\&\;\.\#\n\r\p{Pd}\p{Ps}\p{Pe}\p{Pc}\p{N}\p{L}\p{Sc}\p{Z}]{0,10000}$`

I removed the unnecessary backslashes, corrected the unicode matching groups to actually reference what was intended, and moved the minus sign from the middle of the regex (creating a range) and at the very beginning.  Thanks to @kwwall for showing me that trick.  

As to why that unit test was failing before:  I know there were some minor tweaks made to Java's regex engine between 8 and 12, but I've tested against 8, and 12-14 and that regex is fine.  I can't explain why the original programmer thought why THAT input string ought to fail other than they were working with an already broken regex.  It should pass.

**Comment by DavidSmithAvaloq** [comment]
I am sorry that I confused things by copying and pasting the string before it had been loaded - our test code appears to collapse the double backslashes before passing to esapi - I guess what I copied is used to initialize a Java string. When I ran the test with a debugger, it appears that our validation sting is actually:
```^[+>=<?!%_,~:@|\-\/\*\&\;\.\#\n\r\p{Pd}\p{Ps}\p{Pe}\p{Pc}\p{N}\p{L}\p{Sc}\p{Z}]{0,10000}$``` which is close to what you suggest. I think because the minus sign is preceded by a backslash it loses its usual special status.

The HTMLEntityCodec gets used when the string is canonicalized and so the input string ```&quot;test&quot;``` becomes ```"test"``` and quotes are not allowed by the validation string.

**Comment by xeno6696** [comment]
I am unable to reproduce.  

![image](https://user-images.githubusercontent.com/9502785/106927744-b40dd180-66cf-11eb-997c-927006c621cb.png)

input `"test"`

Results in a ValidationException relating to the regex and a return of "false" from "Validator.isValidInput( args )"

Here's what I wrote in the ESAPI unit test:  

```
public void testavaloqLooseSafeString(){
    	Validator v = ESAPI.validator();
    	boolean isValid = v.isValidInput("RegexString", "&quot;test&quot;", "avaloqLooseSafeString", 2147483647, true, true);
    	assertFalse(isValid);
    }
```
with this in validation.properites:

`Validator.avaloqLooseSafeString=^[+>=<?!%_,~:@|\\-\\/\\*\\&\\;\\.\\#\\n\\r\\p{Pd}\\p{Ps}\\p{Pe}\\p{Pc}\\p{N}\\p{L}\\p{Sc}\\p{Z}]{0,10000}$
`

**Comment by DavidSmithAvaloq** [comment]
Which version are you using for the test? If I look at the current version of StringValidationRule I can see that getValid calls checkWhitelist(context, input); which in turn will call checkWhitelist(context, input, input); which means checkWhitelist must be called with input and orig as the same value, but in your screenshot checkWhitelist has clearly been called with two different values. I cannot see how that is possible with the current version, but I suppose I'd need to see the stack trace to understand how you got there. I am currently back using the older version of the library - tomorrow I will set up a new test with the new version of the library and send more details - I will include some screenshots.

**Comment by xeno6696** [comment]
@DavidSmithAvaloq I don't know what to say... @kwwall will maybe want to remove me from the project after this lol.... 

So... the version of the file that I have doesn't apparently exist anywhere in the repository history.  What's _really_ weird, is that I **_don't remember_** working on this file in the last few months.  But apparently I didn't leave myself a note and it seems that the fix is *precisely* to fix this issue.  In trying to reclaim my sanity I found #517 which appears to have the same root cause though I didn't have a commit there to save my work.

**Comment by kwwall** [comment]
@xeno6696 - Ha! Sorry, you can't get away that easily. Good help is hard to find. I take it you maybe were working in 2 different directories or on 2 different hosts where you made the commit and did the push from one and on the other, never followed up with a 'git pull'. I've done that before once or twice myself across 2 different laptops so I learned to stop doing that. Mostly it was only when I was setting my newer, much faster laptop and didn't have all the software development tools installed on it yet, so I was going back and forth between. I probably should have set up an NFS share between then or something like that. Now that my new laptop is set up, I stopped doing that.

Anyhow, do you have this worked out where you can create a PR or what? I'd try to give you an hand but I'm trying to get ESAPI to work with the new AntiSamy 1.6.0-SNAPSHOT. They are planning on doing the 1.6.0 release RSN.

**Comment by xeno6696** [comment]
@kwwall @jeremiahjstacey,

I have both issues mostly squashed but I have a failing unit test that seems to tickle my memory:
`assertFalse(instance.isValidInput("test", "jeff\\WILLIAMS", "HTTPParameterValue", 100, false));`  (line 546 in ValidatorTest)

Of course in java Stringspeak the literal value at runtime is `jeff\WILLIAMS`.  The PercentCodec _seems_ to have an error here when it runs during the canonicalize operation, where it gets triggered on `\` and then looks ahead, in this case to see `W` but then instead of leaving the string alone (because it's not finding an escape) we append the `W`.  I remember sometime in the last six months Jeremiah working an issue related to the `PercentCodec`.  Sounds similar.  

@kwwall  I just saw your note.  If you're OK with leaving this particular unit test in a fail mode, I can commit and PR what I've got, from what I can see there are no other failures related to this change.

**Comment by xeno6696** [comment]
@kwwall or @jeremiahjstacey 

Please test this for me as I haven't built the whole project in long enough of a time that I can't even get it to test mode:

`StringValidationRule.getValid(String, String):`

```
	public String getValid( String context, String input ) throws ValidationException
	{
		String data = null;

		// check for empty/null
		if(checkEmpty(context, input) == null)
			return null;

		// check length
		checkLength(context, input);

		// canonicalize
		if (canonicalizeInput) {
		    data = encoder.canonicalize(input);
		} else {
		    String message = String.format("Input validation excludes canonicalization.  Context: %s   Input: %s", context, input);
		    LOGGER.warning(Logger.SECURITY_AUDIT, message);
            data = input;
		}

		// check whitelist patterns
		checkWhitelist(context, data, input);

		// check blacklist patterns
		checkBlacklist(context, data, input);
			
		// validation passed
		return data;
	}
```

Please tell me whether or not ValidatorTest fails when you run with this change, and if so, which unit tests.

**Comment by xeno6696** [comment]
For those reading along but not playing, there have been some build failures associated with some more recent project updates and I'm not up to date yet as to what I've got to do to resolve the build issues yet as they're not committed.  Or if they are, my local history is jacked up and I need a clean clone.

**Comment by kwwall** [comment]
Sometimes the best advice is "burn it all to the ground and start over". But if you don't want to do that because you have local changes, you at least need to do a 'git pull' or a 'git fetch' and selectively merge stuff. Usually I opt for the former, but maybe that's just me.

**Comment by DavidSmithAvaloq** [comment]
@xeno6696 thank you for sticking with the problem until we understood it.

**Comment by kwwall** [comment]
@xeno6696 -- What is the status on this? Looks like you maybe fixed it in your fork, but then never made a PR for it??!?!?! See https://github.com/xeno6696/esapi-java-legacy/commit/9a84f831460c8bfc393b86d4d96c65f32c074aca for details.

If this is fixed, I'm okay with delaying another day or two so you can verify this fixed and get me a PR to include in the release.

**Comment by xeno6696** [comment]
Fixed.  ready to go.

**Comment by kwwall** [comment]
@xeno6696 - Thanks, I will merge later after work, but could you change the PR title to something a bit more descriptive than "Sorry I'm late"... Like "Fix for issue 588" or whatever so it looks a bit more descriptive if people pull it up. Thanks.

**Comment by kwwall** [comment]
Closed via PR #611
